pkg/chains/formats/slsa/v2alpha2/internal/pipelinerun/pipelinerun.go

```diff
@@ -0,0 +1,112 @@
+*/
+
+package pipelinerun
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v1"
+	"github.com/tektoncd/chains/pkg/chains/formats/slsa/extract"
+	resolveddependencies "github.com/tektoncd/chains/pkg/chains/formats/slsa/v2alpha2/internal/resolved_dependencies"
+	"github.com/tektoncd/chains/pkg/chains/objects"
+)
+
+const pipelineRunResults = "pipelineRunResults/%s"
+
+// GenerateAttestation generates a provenance statement with SLSA v1.0 predicate for a pipeline run.
+func GenerateAttestation(ctx context.Context, builderID string, pro *objects.PipelineRunObject) (interface{}, error) {
+	rd, err := resolveddependencies.PipelineRun(ctx, pro)
+	if err != nil {
+		return nil, err
+	}
+	bp, err := byProducts(pro)
+	if err != nil {
+		return nil, err
+	}
+	att := intoto.ProvenanceStatementSLSA1{
+		StatementHeader: intoto.StatementHeader{
+			Type:          intoto.StatementInTotoV01,
+			PredicateType: slsa.PredicateSLSAProvenance,
+			Subject:       extract.SubjectDigests(ctx, pro),
+		},
+		Predicate: slsa.ProvenancePredicate{
+			BuildDefinition: slsa.ProvenanceBuildDefinition{
+				BuildType:            "https://tekton.dev/chains/v2/slsa",
+				ExternalParameters:   externalParameters(pro),
+				InternalParameters:   internalParameters(pro),
+				ResolvedDependencies: rd,
+			},
+			RunDetails: slsa.ProvenanceRunDetails{
+				Builder: slsa.Builder{
+					ID: builderID,
+				},
+				BuildMetadata: metadata(pro),
+				Byproducts:    bp,
+			},
+		},
+	}
+	return att, nil
+}
+
+func metadata(pro *objects.PipelineRunObject) slsa.BuildMetadata {
+	m := slsa.BuildMetadata{
+		InvocationID: string(pro.ObjectMeta.UID),
+	}
+	if pro.Status.StartTime != nil {
+		utc := pro.Status.StartTime.Time.UTC()
+		m.StartedOn = &utc
+	}
+	if pro.Status.CompletionTime != nil {
+		utc := pro.Status.CompletionTime.Time.UTC()
+		m.FinishedOn = &utc
+	}
+	return m
+}
+
+// internalParameters adds the tekton feature flags that were enabled
+// for the pipelinerun.
+func internalParameters(pro *objects.PipelineRunObject) map[string]any {
+	internalParams := make(map[string]any)
+	if pro.Status.Provenance != nil && pro.Status.Provenance.FeatureFlags != nil {
+		internalParams["tekton-pipelines-feature-flags"] = *pro.Status.Provenance.FeatureFlags
+	}
+	return internalParams
+}
+
+// externalParameters adds the pipeline run spec
+func externalParameters(pro *objects.PipelineRunObject) map[string]any {
+	externalParams := make(map[string]any)
+	externalParams["runSpec"] = pro.Spec
+	return externalParams
+}
+
+// byProducts contains the pipelineRunResults
+func byProducts(pro *objects.PipelineRunObject) ([]slsa.ResourceDescriptor, error) {
+	byProd := []slsa.ResourceDescriptor{}
+	for _, key := range pro.Status.PipelineResults {
+		content, err := json.Marshal(key.Value)
+		if err != nil {
+			return nil, err
+		}
+		bp := slsa.ResourceDescriptor{
```

Comment:
Maybe set the `MediaType` to `application/json`?

Reply:
What about setting `Digest`? It's not required if `Content` is set, but maybe there's _some_ value?

Reply:
I don't know what I would set it to. Do you have some thoughts?

Reply:
It would be the Digest of the Content. You know... that seems sort of silly. Let's leave Digest out for now.

Reply:
SG!